### ISSUE TITLE
Unexpected behaviour in Bun's static analysis.

### ISSUE THREAD

**Issue by vdata1** [oven-sh/bun#13992]
### What version of Bun is running?

1.1.28

### What platform is your computer?

Darwin 23.6.0 x86_64 i386

### What steps can reproduce the bug?

Hi, 

I want to report a couple of peculiar, possibly non-standard Bun behaviour based on the attached four code snippets. 

It seems to me that Bun's static analysis is too aggressive in the pre-execution step and considers dead parts in code during the analysis. By looking at the first code snippet, Bun throws an error at line 3 after the `return` statement, in unreachable code. Moreover, in code example number 2, the dead code is constructed based on logical irreachability, the code block inside `function b(){}` is not reachable, and still, Bun gives the same error as the first code snippet. Moreover, in code example 3, even though the function a() is never invoked, Bun throws an error saying 'obj' is a constant and should not be modified. The fourth code shows that the static analysis also struggles with the 'with' construct: even though the code modifies a mutable property of 'obj', Bun throws saying again that 'obj' is a constant, which it is not in this case. We cross-checked all these examples with Node.js and Deno and we did not observe any of the reported behavior in those runtimes. 

```javascript
function b(){
    return;
    obj = 2;
}
const obj = 1;
b();
```

```javascript
function b() {    
   if (false) {
    obj = 2;
   }        
}
const obj = 1;        
b();
```

```javascript
"use strict";
function a() {
  obj = 2;
}
const obj = 1;
```

```javascript 
const obj = { obj: 2 };

with (obj) {
    obj=10;
};
console.log(obj);
```

**Comment by Jarred-Sumner** [comment]
yeah, the const reassignment error shouldn't be triggering when not in strict mode
